xarray/indexes/interval_index.py

```diff
@@ -0,0 +1,185 @@
+            elif v.ndim == 1:
+                dim, _ = k, v
+
+        bounds = bounds.transpose(..., dim)
+        left, right = bounds.data.tolist()
+        # TODO: support non-dimension coordinates (pass variable name to pd.IntervalIndex.from_arrays)
+        # TODO: propagate coordinate dtype (pass it to PandasIndex constructor)
+        # TODO: add "closed" build option (maybe choose "closed='both'" as default here? to be consistent with
+        # CF conventions: https://cfconventions.org/cf-conventions/cf-conventions.html#bounds-one-d)
+        index = PandasIndex(pd.IntervalIndex.from_arrays(left, right), dim)
+        bounds_dim = (set(bounds.dims) - set(dim)).pop()
+
+        return cls(index, bounds_name, str(bounds_dim))
+
+    @classmethod
+    def concat(
+        cls,
+        indexes: Sequence[IntervalIndex],
+        dim: Hashable,
+        positions: Iterable[Iterable[int]] | None = None,
+    ) -> IntervalIndex:
+        new_index = PandasIndex.concat(
+            [idx._index for idx in indexes], dim, positions=positions
+        )
+
+        if indexes:
+            bounds_name = indexes[0]._bounds_name
+            bounds_dim = indexes[0]._bounds_dim
+            if any(
+                idx._bounds_name != bounds_name or idx._bounds_dim != bounds_dim
+                for idx in indexes
+            ):
+                raise ValueError(
+                    f"Cannot concatenate along dimension {dim!r} indexes with different "
+                    "boundary coordinate or dimension names"
+                )
+        else:
+            bounds_name = new_index.index.name + "_bounds"
+            bounds_dim = "bnd"
+
+        return cls(new_index, bounds_name, bounds_dim)
+
+    @property
+    def _pd_index(self) -> pd.IntervalIndex:
+        # For typing purpose only
+        # TODO: cleaner to make PandasIndex a generic class, i.e., PandasIndex[pd.IntervalIndex]
+        # will be easier once PEP 696 is fully supported (starting from Python 3.13)
+        return cast(pd.IntervalIndex, self._index.index)
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Any, Variable]:
+        if variables is None:
+            variables = {}
+        empty_var = Variable((), 0)
+        bounds_attrs = variables.get(self._bounds_name, empty_var).attrs
+        mid_attrs = variables.get(self._index.dim, empty_var).attrs
+
+        # TODO: create a PandasIndexingAdapter subclass for the boundary variable
+        # and wrap it here (avoid data copy)
+        bounds_var = Variable(
+            dims=(self._bounds_dim, self._index.dim),
+            data=np.stack([self._pd_index.left, self._pd_index.right], axis=0),
+            attrs=bounds_attrs,
+        )
+        # TODO: use PandasIndexingAdapter directly (avoid data copy)
+        # and/or maybe add an index build option to preserve original labels?
+        # (if those differ from interval midpoints as defined by pd.IntervalIndex)
```

Comment:
I think we should always save the central value and return that.

Reply:
Agreed. This is sub-optimal in the case where the central values exactly correspond to what is returned `pd.IntervalIndex.mid` (e.g., when creating an IntervalIndex directly from a `pd.IntervalIndex`), but as I was trying to account for this special case I went into all sorts of small complications, so I think it'll be easier to always save the central values into a separate `PandasIndex` here.